keps/sig-network/4427-relaxed-dns-search-validation/README.md

```diff
@@ -0,0 +1,473 @@
+
+Introduce a RelaxedDNSSearchValidation feature gate which is disabled by default. When the feature gate is enabled,
+a new DNS name validation function will be used, which keeps the existing validation logic but also allows an underscore (`_`) in any place
+where a dash (`-`) would be allowed currently.
+
+Since the relaxed check allows previously invalid values, care must be taken to support cluster downgrades safely. To accomplish this, the validation will distinguish between new resources and updates to existing resources:
+- When the feature gate is disabled:
+  - (a) New resources will use strict validation based on RFC-1123 (no change to current validation)
+  - (b) Updates to existing resources will use relaxed validation if any search string in the existing list fails strict validation
+- When the feature gate is enabled:
+  - (c) New resources will use relaxed validation.
+  - (d) Updates to existing resources will use relaxed validation.
+
+This means that it is safe to downgrade a cluster with the feature gate enabled to a version where the feature gate is present (whether it’s enabled or disabled). It is not safe, in general, to downgrade from a cluster with the gate enabled to a version prior to the gate being introduced, since values may have been written to storage which will no longer pass validation. However, this scenario requires opting in through enabling the gate. In practice, the recommended approach would be to only enable to the gate after upgrading from a version with relaxed checking already present.
+
+As long as the gate is disabled, there is no compatibility change, so cluster downgrades are not affected by the feature.
+
+### Risks and Mitigations
+
+The change is opt-in, since it requires configuring a search string with an underscore. So there is no risk beyond
+the upgrade/downgrade risks which are addressed in the Proposal section.
+
+## Design Details
+
+See Proposal
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+The goal is to ensure that we don't accept enhancements with inadequate testing.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+<!--
+Based on reviewers feedback describe what additional tests need to be added prior
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+##### Unit tests
+
+Added validation will be covered by unit tests along with unit tests covering the behavior
+when the gate is enabled or disabled.
+
+##### Integration tests
```

Comment:
This is where the most important tests should be written. The apiserver can be started in integration with feature flags set and we have confidence that those tests mirror what will happen in production. Let's list out the test cases we need to validate the ratcheting here:

- feature gate on
- feature gate off
- feature gate on, name with_ written, feature gate off, updates still work that change other fields or that change name to something without _, (what if the name is changed to a new value with _?)
- ...